### ISSUE TITLE
Msf::DBManager tests failing

### ISSUE THREAD

**Issue by itsbriany** [rapid7/metasploit-framework#7555]
Hi MSF community!
 
So far, I have successfully set up my dev environment and MSF is working great!

However, I have ran into an issue where all of the `Msf::DBManager` tests fail. This probably means that I have not properly set up postgres for the test environment (all the other tests pass).

## Steps to reproduce

After following the [documentation](https://github.com/rapid7/metasploit-framework/wiki/Setting-Up-a-Metasploit-Development-Environment) for setting up the dev environment, running `msfconsole -qx "db_status; exit"` returns `[*] postgresql connected to msf_dev_db` (this tells me that MSF is working fine with the development database). Currently, I have both `msf_dev_db` and `msf_test_db` owned by `msfdev`. Also, my `$HOME/.msf4/database.yml` file is currently the following:

```yml
# Development Database
development: &pgsql
  adapter: postgresql
  database: msf_dev_db
  username: msfdev
  password: msfdev
  host: localhost
  port: 5432
  pool: 5
  timeout: 5

# Production database -- same as dev
production: &production
  <<: *pgsql

# Test database -- not the same, since it gets dropped all the time
test:
  <<: *pgsql
  database: msf_test_db
```

## Expected behavior
All tests should pass when running
```bash
rspec spec/lib/msf/db_manager_spec.rb
```

## Current behavior
All tests fail when running
```bash
rspec spec/lib/msf/db_manager_spec.rb
```

All the errors are pretty much the same, but what stands out to me are the following:

```
CONTENT SNIPPED...

5) Msf::DBManager it should behave like Msf::DBManager::Adapter
     Failure/Error: ::Mdm::Workspace.default

     ActiveRecord::StatementInvalid:
       PG::UndefinedTable: ERROR:  relation "workspaces" does not exist
       LINE 5:                WHERE a.attrelid = '"workspaces"'::regclass
                                                 ^
       :               SELECT a.attname, format_type(a.atttypid, a.atttypmod),
                            pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod
                       FROM pg_attribute a LEFT JOIN pg_attrdef d
                         ON a.attrelid = d.adrelid AND a.attnum = d.adnum
                      WHERE a.attrelid = '"workspaces"'::regclass
                        AND a.attnum > 0 AND NOT a.attisdropped
                      ORDER BY a.attnum
     Shared Example Group: "Msf::DBManager::Adapter" called from ./spec/lib/msf/db_manager_spec.rb:21

CONTENT SNIPPED...

# ./lib/msf/core/db_manager/workspace.rb:13:in `block in default_workspace'
# /Users/itsbriany/.rvm/gems/ruby-2.3.1@metasploit-framework/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract/connection_pool.rb:
292:in `with_connection'
# ./lib/msf/core/db_manager/workspace.rb:12:in `default_workspace'
# ./spec/support/shared/contexts/msf/db_manager.rb:14:in `block (2 levels) in <top (required)>'

CONTENT SNIPPED...
```

It looks like MSF cannot find a default workspace for `msf_test_db`. However, I would assume that default workspaces should be automatically created given that a proper connection with the database has been established?

### Metasploit version

Framework: 4.12.41-dev-77bacac
Console  : 4.12.41-dev-77bacac

### I installed Metasploit with:
Source install
```bash
ruby -v
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin16]
```

### OS

**MacOS 10.12.1 (Sierra)**

## Update
Based on the error, it looks like the tests cannot find the MSF default workspace (although when connecting to `msf_dev_db`, there IS a default workspace). Therefore, I got most `Msf::DBManager` tests to pass after configuring `$HOME/.msf4/databases.yml` to temporarily point the development database to `msf_test_db` (instead of `msf_dev_db`), and then running `msfconsole`. After that, I made some changes to `$HOME/.msf4/databases.yml`:

```yml
# Development Database
development: &pgsql
  adapter: postgresql
  database: msf_dev_db
  username: msfdev
  password: msfdev
  host: localhost
  port: 5432
  pool: 5
  timeout: 5

# Production database -- same as dev
production: &production
  <<: *pgsql

# Test database -- not the same, since it gets dropped all the time
test:
  <<: *pgsql
  database: msf_test_db
  username: msfdev
  password: msfdev
  host: localhost
  port: 5432
```

However, I do not think that my solution was the proper fix because four tests still fail:

```bash
  1) Msf::DBManager it should behave like Msf::DBManager::ModuleCache #update_all_module_details with migrated without modules_caching with Mdm::Module::Detai
ls #ready true with existing Mdm::Module::Detail#file without same Mdm::Module::Detail#mtime and File.mtime it should behave like Msf::DBManager#update_all_mo
dule_details refresh should destroy Mdm::Module::Detail
     Failure/Error:
       expect {
         update_all_module_details
       }.to change(Mdm::Module::Detail, :count).by(-1)

       expected #count to have changed by -1, but was changed by -2
     Shared Example Group: "Msf::DBManager#update_all_module_details refresh" called from ./spec/support/shared/examples/msf/db_manager/module_cache.rb:706
     Shared Example Group: "Msf::DBManager::ModuleCache" called from ./spec/lib/msf/db_manager_spec.rb:35

CONTENT SNIPPED...

2) Msf::DBManager it should behave like Msf::DBManager::ModuleCache #update_all_module_details with migrated without modules_caching with Mdm::Module::Detai
ls #ready false it should behave like Msf::DBManager#update_all_module_details refresh should destroy Mdm::Module::Detail
     Failure/Error:
       expect {
         update_all_module_details
       }.to change(Mdm::Module::Detail, :count).by(-1)

       expected #count to have changed by -1, but was changed by -2
     Shared Example Group: "Msf::DBManager#update_all_module_details refresh" called from ./spec/support/shared/examples/msf/db_manager/module_cache.rb:682
     Shared Example Group: "Msf::DBManager::ModuleCache" called from ./spec/lib/msf/db_manager_spec.rb:35

CONTENT SNIPPED...

 3) Msf::DBManager it should behave like Msf::DBManager::ModuleCache #purge_all_module_details with migrated without modules_caching should destroy all Mdm::
Module::Details
     Failure/Error:
       expect {
         purge_all_module_details
       }.to change(Mdm::Module::Detail, :count).by(-module_detail_count)

       expected #count to have changed by -2, but was changed by -3
     Shared Example Group: "Msf::DBManager::ModuleCache" called from ./spec/lib/msf/db_manager_spec.rb:35

CONTENT SNIPPED...

 4) Msf::DBManager it should behave like Msf::DBManager::ModuleCache #search_modules with author keyword with Mdm::Module::Author#email should match Mdm::Mod
ule::Author#email
     Failure/Error:
       expect(
         module_details.all? { |module_detail|
           module_detail.authors.any? { |module_author|
             module_author.email == target_module_author.email
           }
         }
       ).to eq true

       expected: true
            got: false

       (compared using ==)
     Shared Example Group: "Msf::DBManager::ModuleCache" called from ./spec/lib/msf/db_manager_spec.rb:35
```

So in a nutshell, what is the simplest way to install MSF and get all of the `Msf::DBManager` tests to pass? :)

**Comment by itsbriany** [comment]
So I figured out how to fix this and why the tests were failing:

When `msf_test_db` is first created, it has no tables. It appears that `msfconsole` automatically generates the tables for the dev environment (commonly referred to as database migrations) on startup (this is why most tests previously passed when temporarily pointing the dev environment to `msf_test_db` in `$HOME/.msf4/database.yml`). Invoking `rake` to run the tests does not automatically migrate the database schemas for the test environment. Hence when the tests run, they cannot find their tables.

Executing the following command resolves this issue:
```bash
rake db:migrate RAILS_ENV=test
```

After that, all tests passed.

Perhaps, this should be mentioned in the [documentation](https://github.com/rapid7/metasploit-framework/wiki/Setting-Up-a-Metasploit-Development-Environment) for anyone willing to run tests after setting up a dev environment?

**Comment by darkbushido** [comment]
You should run ```rake db:test:prepare``` before running ```rspec spec/lib/msf/db_manager_spec.rb``` 

I'm going to create a pull request that will let you run ```rake spec SPEC=spec/lib/msf/db_manager_spec.rb``` after the change it will auto run rake db:test:prepare

**Comment by vmarquet** [comment]
Could the documentation please be modified to mention the `rake db:test:prepare` command?

This is not specified anywhere and everyone following the documentation to the letter will be stuck at "You should see over 9000 tests run, mostly resulting in green dots, a few in yellow stars, and no red errors.", which won't be true. Given the verbose output of the failing tests, this issue and the solution are not easy to find through Google.

Should I open a new issue?

**Comment by busterb** [comment]
Yeah, this bugs me too. Let's reopen this.

**Comment by busterb** [comment]
Strangely, I'm almost positive that at some point in the past, one merely had to run 'rake spec' and this would work automatically. I wonder if the order of tests got somehow changed implicitly along the way, causing it to start failing.
